libs/sync/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+//! Only synchronisation primitive implementations should depend on this crate.
+//!
+//! If a crate uses a synchronisation primitive, it should depend on one of the
+//! following:
+//! - `sync_spin`
+//! - `sync_preemption`
+//! - `sync_irq`
+//! - `sync_block`
+
+#![no_std]
+
+pub mod mutex;
+
+pub use mutex::{Mutex, MutexGuard};
+
+/// A synchronisation flavour.
+pub trait Flavour {
+    /// Initial value for the lock data.
+    const INIT: Self::LockData;
+
+    /// Additional data stored on the lock.
+    type LockData;
+
+    type DeadlockPrevention: DeadlockPrevention;
+
+    /// Acquires the given mutex.
+    fn mutex_lock<'a, T>(
+        mutex: &'a mutex::SpinMutex<Self::DeadlockPrevention, T>,
+        data: &'a Self::LockData,
+    ) -> mutex::SpinMutexGuard<'a, Self::DeadlockPrevention, T>
+    where
+        Self: Sized;
+
+    /// Performs any necessary actions after unlocking the mutex.
+    fn post_unlock(mutex: &Self::LockData)
+    where
+        Self: Sized;
+}
+
+/// A deadlock prevention method.
+pub trait DeadlockPrevention {
```

Comment:
I generally like this name, but unfortunately it's a bit too narrow in scope to be correct -- we implement it for things that have no bearing on deadlock, e.g., blocking a task.

It's more like a pre/post lock action catch-all trait. We can bikeshed the name a bit, either here or on discord.